04-onion-routing.md

```diff
@@ -367,9 +378,60 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Trampoline Payments
+
+Trampoline payments allow nodes with an incomplete view of the network to
+delegate the construction of parts of the route to trampoline nodes.
+
+The origin node only needs to select a set of trampoline nodes and to know a
+route to the first trampoline node. Each trampoline node is responsible for
+finding its own route to the next trampoline node. The last trampoline node
+must be the final recipient.
+
+The `trampoline_onion_packet` has a variable size to allow implementations to
+choose their own trade-off between flexibility and privacy. It's recommended to
+add trailing filler data to the `trampoline_onion_packet` when using a small
+number of hops. It uses the same onion construction as the `onion_packet` and
+is embedded inside an `onion_packet`.
+
+Trampoline nodes are free to use as many hops as they want between themselves
+as long as they are able to create a route that satisfies the `cltv` and `fees`
+requirements contained in the onion.
+
+#### Requirements
+
+A sending node:
+
+- If the invoice doesn't support the `trampoline_routing` feature:
+  - MUST NOT use trampoline routing to pay that invoice
+- MUST ensure that each hop in the `trampoline_onion_packet` supports `trampoline_routing`
+- MUST encrypt the `trampoline_onion_packet` with the same construction as `onion_packet`
+- MAY add trailing filler data similar to what is done in the `onion_packet`
+- MUST use a different `session_key` for the `trampoline_onion_packet` and the `onion_packet`
+- MUST include the `trampoline_onion_packet` tlv in the _last_ hop's payload of the `onion_packet`
+- MUST include the invoice's `payment_secret` in the _last_ hop's payload of the `trampoline_onion_packet`
+- MUST generate a different `payment_secret` to use in the outer onion
+
+When processing a `trampoline_onion_packet`, a receiving node:
+
+- If it doesn't support `trampoline_routing`:
+  - MUST report a route failure to the origin node
+- Otherwise, if it supports `trampoline_routing`:
+  - MUST process the `trampoline_onion_packet` as an `onion_packet`
+  - MUST fail the HTLC if dictated by the requirements under [Failure Messages](#failure-messages)
+  - MUST compute a route to the next trampoline node
+  - MUST include the peeled `trampoline_onion_packet` in the `hop_payload` for the next trampoline node
+  - MAY aggregate the incoming multi-part payment before forwarding
+  - If it uses a multi-part payment to forward to the next trampoline node:
+    - MUST generate a new `payment_secret` to use in the outer onion
```

Comment:
payment_secret is surely only used by the final recipient?  Nobody else can check it anyway, can they?

Reply:
The invoice one is put in the trampoline onion to be read only by the final recipient.
But when you're constructing a new MPP set between trampoline T1 and T2, it makes sense to use a `payment_secret` for that MPP flow, and it must be generated by `T1` so it can only be in the outer onion.